tests/unit/test_charm.py

```diff
@@ -18,129 +19,152 @@
+                }
+            ),
+        )
+
+        self.harness.set_leader(True)
+        self.harness.set_can_connect("lego", True)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+        self.r_id = self.harness.add_relation("certificates", "remote")
+        self.harness.add_relation_unit(self.r_id, "remote/0")
+
+    def test_given_empty_pebble_plan_when_pebble_ready_then_status_is_active(self):
+        initial_plan = self.harness.get_container_pebble_plan("lego")
+        self.assertEqual(initial_plan.to_yaml(), "{}\n")
+        expected_plan = {}
+        container = self.harness.model.unit.get_container("lego")
+        self.harness.charm.on.lego_pebble_ready.emit(container)
+        updated_plan = self.harness.get_container_pebble_plan("lego").to_dict()
+        self.assertEqual(expected_plan, updated_plan)
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+
+    @patch(
+        f"{TLS_LIB_PATH}.TLSCertificatesProvidesV1.set_relation_certificate",
     )
-
-    request_cert(harness)
-
-
-def test_failing_request(harness):
-    harness._backend._pebble_clients["lego"].exec = partial(
-        check_exec_args,
-        harness,
-        Mock(**{"wait_output.side_effect": ExecError("lego", 1, "barf", "rip")}),
-    )
-    harness._backend._pebble_clients["lego"].push(
-        "/tmp/.lego/certificates/foo.crt", source=test_lego.read_bytes(), make_dirs=True
-    )
-
-    request_cert(harness)
-    assert harness.charm.unit.status == BlockedStatus(
-        "Error getting certificate. Check logs for details"
-    )
-
-
-def test_cannot_connect(harness):
-    harness.set_can_connect("lego", False)
-    request_cert(harness)
-    assert harness.charm.unit.status == WaitingStatus("Waiting for container to be ready")
+    def test_given_cmd_when_certificate_creation_request_then_certificate_is_set_in_relation(
+        self, mock_set_relation_certificate
+    ):
+        self.harness._backend._pebble_clients["lego"].exec = partial(
+            self.check_exec_args, self.harness, Mock(wait_output=lambda: (None, None))
+        )
+        self.harness._backend._pebble_clients["lego"].push(
+            "/tmp/.lego/certificates/foo.crt", source=test_cert.read_bytes(), make_dirs=True
+        )
+
+        csr = self.request_cert()
+        with open(test_cert, "r") as file:
+            expected_certs = (file.read()).split("\n\n")
+        mock_set_relation_certificate.assert_called_with(
+            certificate=expected_certs[0],
+            certificate_signing_request=csr,
+            ca=expected_certs[-1],
+            chain=list(reversed(expected_certs)),
+            relation_id=self.r_id,
+        )
+
+    def test_given_command_execution_fails_when_certificate_creation_request_then_request_fails_and_status_is_blocked(
+        self,
+    ):
+        self.harness._backend._pebble_clients["lego"].exec = partial(
```

Comment:
Humm we shouldn't have to call the harness private method. First the call could change, second the call is confusing.

Reply:
✅